hexl/include/hexl/fft/fft.hpp

```diff
@@ -0,0 +1,150 @@
+                    std::move(a), std::forward<AllocatorArgs>(args)...))) {}
+
+  /// @brief Compute forward FFT. Results are bit-reversed.
+  /// @param[out] result Stores the result
+  /// @param[in] operand Data on which to compute the FFT
+  /// @param[in] in_scale Scale applied to output values
+  void ComputeForwardFFT(std::complex<double_t>* result,
+                         const std::complex<double_t>* operand,
+                         const double_t* in_scale = nullptr);
+
+  /// @brief Compute inverse FFT. Results are bit-reversed.
+  /// @param[out] result Stores the result
+  /// @param[in] operand Data on which to compute the FFT
+  /// @param[in] in_scale Scale applied to output values
+  void ComputeInverseFFT(std::complex<double_t>* result,
+                         const std::complex<double_t>* operand,
+                         const double_t* in_scale = nullptr);
+
+  /// @brief Construct floating-point values from CRT-composed polynomial with
+  /// integer coefficients.
+  /// @param[out] res Stores the result
+  /// @param[in] plain Plaintext
+  /// @param[in] threshold Upper half threshold with respect to the total
+  /// coefficient modulus
+  /// @param[in] decryption_modulus Product of all primes in the coefficient
+  /// modulus
+  /// @param[in] inv_scale Scale applied to output values
+  /// @param[in] mod_size Size of coefficient modulus parameter
+  /// @param[in] coeff_count Degree of the polynomial modulus parameter
+  void BuildFloatingPoints(std::complex<double_t>* res, const uint64_t* plain,
+                           const uint64_t* threshold,
+                           const uint64_t* decryption_modulus,
+                           const double_t inv_scale, size_t mod_size,
+                           size_t coeff_count);
+
+  /// @brief Returns the root of unity power at bit-reversed index i.
+  /// @param[in] i Index
+  std::complex<double_t> GetComplexRootOfUnity(size_t i) {
+    return GetComplexRootsOfUnity()[i];
+  }
+
+  /// @brief Returns the root of unity in bit-reversed order
+  const AlignedVector64<std::complex<double_t>>& GetComplexRootsOfUnity()
+      const {
+    return m_complex_roots_of_unity;
+  }
+
+  /// @brief Returns the root of unity power at bit-reversed index i.
+  /// @param[in] i Index
+  std::complex<double_t> GetInvComplexRootOfUnity(size_t i) {
+    return GetInvComplexRootsOfUnity()[i];
+  }
+
+  /// @brief Returns the inverse root of unity in bit-reversed order
+  const AlignedVector64<std::complex<double_t>>& GetInvComplexRootsOfUnity()
+      const {
+    return m_inv_complex_roots_of_unity;
+  }
+
+  /// @brief Returns the degree N
+  uint64_t GetDegree() const { return m_degree; }
+
+ private:
+  // Computes 1~(n-1)-th powers and inv powers of the primitive 2n-th root
+  void ComputeComplexRootsOfUnity();
+
+  // PI value used to calculate the roots of unity
+  static constexpr double PI_ = 3.1415926535897932384626433832795028842;
```

Comment:
I have only seen this used in one place. Could it not be defined where it is used?

Reply:
Moved